# Gate 0-1 Drill C: 숫자 label 함정

목표:
- 숫자로 저장된 y가 회귀 target인지, class label encoding인지 구분한다.
- X/y, feature/target, 문제 유형을 정확히 판별한다.
- 회귀로 잘못 풀었을 때 생기는 문제를 설명한다.

## 판단 기준

각 문제에서 다음 네 가지를 판단한다.

1. X는 무엇인가?
2. y는 무엇인가?
3. y의 숫자는 크기/거리 의미가 있는가?
4. y가 class 이름을 숫자로 바꾼 것인가?

최종 문제 유형:
- 회귀: y가 연속값, 수치량, 단위가 있는 값
- 이진분류: y가 두 class 중 하나
- 다중분류: y가 여러 class 중 하나
- 순서형 분류: class 사이에 순서는 있으나 거리 간격은 애매한 경우

## C-1. Iris 품종 encoding

Iris 품종을 다음처럼 저장했다.

- 0 = setosa
- 1 = versicolor
- 2 = virginica

꽃의 네 가지 측정값으로 이 값을 예측한다.

답할 것:
- X =
- y =
- feature =
- target =
- 이 숫자는 크기/거리 의미가 있는가?
- class 이름을 숫자로 바꾼 것인가?
- 문제 유형 =
- 회귀로 풀면 생기는 문제 =

In [ ]:
# 힌트:
# 1. sklearn.datasets에서 iris 데이터 불러오기
# 2. iris.data는 X 후보
# 3. iris.target은 y 후보
# 4. iris.feature_names로 column 이름 확인
# 5. iris.target_names로 0,1,2가 어떤 class 이름인지 확인

# 사용해볼 함수/속성:
# load_iris()
# iris.data
# iris.target
# iris.feature_names
# iris.target_names
# pd.DataFrame(...)

from sklearn.datasets import load_iris
import pandas as pd

iris = load_iris()

X_df.head(10)          # X 앞 10행
y_s.head(10)   

print("X shape:", X.shape)        # (150, 4) → 150행, 4열 // 
print("y shape:", y.shape)        # (150,)
print("X 열 개수:", X_df.shape[1])
print("X 행 개수:", X_df.shape[0])
print("feature 이름:", iris.feature_names)

X_df.info()
X_df.describe()   

print("target 이름:", iris.target_names)
print("y 고유값:", pd.unique(y))
print("클래스별 개수:")
print(y_s.value_counts().sort_index())

df = X_df.copy()
df["species_code"] = y_s
df.head()

df["species_name"] = iris.target_names[y]
df.head()

X shape: (8, 6)
y shape: (8,)
X 열 개수: 6
X 행 개수: 8
feature 이름: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
<class 'pandas.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   feature_1_vehicle_capacity_ratio  8 non-null      float64
 1   feature_2_time_day_weight_ratio   8 non-null      float64
 2   feature_3_speed_road_ratio        8 non-null      float64
 3   feature_4_weather_weight          8 non-null      float64
 4   feature_5_accident_road_ratio     8 non-null      float64
 5   feature_6_vehicle_hour_ratio      8 non-null      float64
dtypes: float64(6)
memory usage: 516.0 bytes
target 이름: ['setosa' 'versicolor' 'virginica']
y 고유값: [0 1 2 3]
클래스별 개수:
congestion_code
0    2
1    2
2    2
3    2
Name: count, dtype: int64


IndexError: index 3 is out of bounds for axis 0 with size 3

### 내 답안 C-1

X =  shape(150,4)
y =  shape(150,)
feature =  'sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)'
target =  'setosa' 'versicolor' 'virginica'

이 숫자는 크기/거리 의미가 있는가? =  x 즉 피처는 모두 연속형 수치로서 비율 등간 척도가 아니므로, 사칙연산이 가능하고 (이거 심층디벨롭) 
class 이름을 숫자로 바꾼 것인가? =  클래스 이름을 숫자로 바꾼게 아님. 
문제 유형 =  다중분류문제.
회귀로 풀면 생기는 문제 = 문제 제정의시 연속형 수치로 반환되는 회귀에 대하여 구간별 급간별로 나눌 시 최종 분류문제로서 풀수 있으나 기본적으로 0,1,2 에 대한 예측이므로 - 분류문제가 맞음. 

근거:
-	sepal length (cm)	sepal width (cm)	petal length (cm)	petal width (cm)	species_code	species_name
0	5.1	3.5	1.4	0.2	0	setosa
1	4.9	3.0	1.4	0.2	0	setosa
2	4.7	3.2	1.3	0.2	0	setosa
3	4.6	3.1	1.5	0.2	0	setosa
4	5.0	3.6	1.4	0.2	0	setosa


## C-2. 고객 만족도 1~5점 예측

### 문제

고객 만족도를 1, 2, 3, 4, 5점으로 입력받았다.  
고객 응답 데이터를 이용해 만족도 점수를 예측한다.

### 답할 것

- X =
- y =
- feature =
- target =
- 이 숫자는 크기/거리 의미가 있는가?
- class 이름을 숫자로 바꾼 것인가?
- 문제 유형 =
- 회귀로 풀면 생기는 문제 또는 분류로 풀면 생기는 문제 =
- 애매하다면 왜 애매한가?

In [ ]:
# 실제 데이터가 없으므로 예시 DataFrame을 직접 만든다고 가정한다.
# 예시 feature 후보:
# 응답 시간, 재방문 횟수, 불만 키워드 수, 구매 금액, 상담 횟수
# target 후보:
# satisfaction_score

# 생각할 것:
# 1,2,3,4,5 사이에 순서가 있는가?
# 1점과 2점 차이, 4점과 5점 차이를 같은 거리로 봐도 되는가?
# 회귀, 다중분류, 순서형 분류 중 무엇이 가장 자연스러운가?

from sklearn.datasets import load_만족도
import pandas as pd

만족도 = load_만족도()

X_df.head(10)          # X 앞 10행
y_s.head(10)   

print("X shape:", X.shape)        # (???, ?) → ???행, ?열
print("y shape:", y.shape)        # (???,)
print("X 열 개수:", X_df.shape[1])
print("X 행 개수:", X_df.shape[0])
print("feature 이름:", 만족도.feature_names)

X_df.info()
X_df.describe()   

print("target 이름:", 만족도.target_names)
print("y 고유값:", pd.unique(y))
print("클래스별 개수:")
print(y_s.value_counts().sort_index())

df = X_df.copy()
df["만족도_code"] = y_s
df.head()

df["만족도_name"] = 만족도.target_names[y]
df.head()

### 내 답안 C-2

#코드실행예상

X shape: (???, 5+)
y shape: (???,)
X 열 개수: 응답 시간, 재방문 횟수, 불만 키워드 수, 구매 금액, 상담 횟수 , = 최소 5 이상.
X 행 개수: ???
feature 이름: ['응답 시간 (min)', '재방문 횟수 (int)', '불만 키워드 수 (int)', '구매 금액 (won)','상담 횟수 (int)', +]
<class 'pandas.DataFrame'>
RangeIndex: ??? entries, 0 to ???
Data columns (total 5+ columns):

dtypes: float64(5+)
memory usage: 4.8 KB
target 이름: ['매우불만족' '불만족' '보통' '만족' '매우만족']
y 고유값: [1 2 3 4 5]
클래스별 개수:
species_code
1    
2
3
4
5
Name: count, dtype: int64

X =  
y =  
feature =  ['응답 시간 (min)', '재방문 횟수 (int)', '불만 키워드 수 (int)', '구매 금액 (won)','상담 횟수 (int)',]
target =  '매우불만족' '불만족' '보통' '만족' '매우만족' / 1,2,3,4,5

이 숫자는 크기/거리 의미가 있는가? =  해당 숫자는 상호간 단위가 다르며, 또한 상호 범주형이 아니다. 
class 이름을 숫자로 바꾼 것인가? =  클래스 아이디를 숫자로변환한게 아니다.
문제 유형 =  1~~5중 5중 분류문제.
다른 방식으로 풀 수 있다면? =  재정의후 , 선형회귀로서, 수치를 낸다면 반올림 또는 정수제약 등을 잡아준다면 , 급간에 대하여 연속형(수치형)으로 결과가 나옴에도 분류형(범주형)으로 제시될 수 있을것.

근거:
-

## C-3. 영화 장르 encoding

### 문제

영화 장르를 다음처럼 저장했다.

- 0 = 액션
- 1 = 로맨스
- 2 = 스릴러
- 3 = 다큐멘터리

줄거리 텍스트를 이용해 장르 번호를 예측한다.

### 답할 것

- X =
- y =
- feature =
- target =
- 이 숫자는 크기/거리 의미가 있는가?
- class 이름을 숫자로 바꾼 것인가?
- 문제 유형 =
- 회귀로 풀면 생기는 문제 =

In [ ]:
import pandas as pd
import numpy as np
import re

target_names = np.array(["액션", "로맨스", "스릴러", "다큐멘터리"])
raw_df = pd.DataFrame({
    "plot": [
        "특수부대 요원이 폭발과 총격 액션 속에서 테러리스트를 추격한다.",
        "두 사람이 우연히 만나 사랑과 로맨스를 키우지만 이별 위기에 놓인다.",
        "연쇄 살인범을 쫓는 형사가 어두운 도시에서 스릴러 같은 긴장을 겪는다.",
        "북극곰과 기후 변화를 담은 다큐멘터리 영화가 자연의 아름다움을 보여준다.",
    ],
    "genre_code": [0, 1, 2, 3],
})
keywords = ["액션", "로맨스", "스릴러", "다큐멘터리"]

def extract_keyword_features(text, keywords):
     words = re.findall(r"[가-힣]+", text)
     total_words = max(len(words), 1)
     result = {}
     for kw in keywords:
        count = text.count(kw)
        result[f"{kw}_exists"] = int(count > 0)
        result[f"{kw}_count"] = count
        result[f"{kw}_ratio"] = count / total_words
     return result

feature_df = raw_df["plot"].apply(lambda t: pd.Series(extract_keyword_features(t, keywords)))

X_df = feature_df[[f"{kw}_ratio" for kw in keywords]].copy()
X_df.columns = [f"feature_{i+1}_{kw}" for i, kw in enumerate(keywords)]
X = X_df.values
y = raw_df["genre_code"].values
y_s = pd.Series(y, name="genre_code")
print(X_df)
print(feature_df)
print("target 이름:", target_names)
print("y 고유값:", np.unique(y))
print("클래스별 개수:")
print(y_s.value_counts().sort_index())

print("=== 로데이터 (줄거리 + 장르 code) ===")
print(raw_df.head())





   feature_1_액션  feature_2_로맨스  feature_3_스릴러  feature_4_다큐멘터리
0         0.125            0.0            0.0         0.000000
1         0.000            0.1            0.0         0.000000
2         0.000            0.0            0.1         0.000000
3         0.000            0.0            0.0         0.111111
   액션_exists  액션_count  액션_ratio  로맨스_exists  로맨스_count  로맨스_ratio  \
0        1.0       1.0     0.125         0.0        0.0        0.0   
1        0.0       0.0     0.000         1.0        1.0        0.1   
2        0.0       0.0     0.000         0.0        0.0        0.0   
3        0.0       0.0     0.000         0.0        0.0        0.0   

   스릴러_exists  스릴러_count  스릴러_ratio  다큐멘터리_exists  다큐멘터리_count  다큐멘터리_ratio  
0         0.0        0.0        0.0           0.0          0.0     0.000000  
1         0.0        0.0        0.0           0.0          0.0     0.000000  
2         1.0        1.0        0.1           0.0          0.0     0.000000  
3         0.0        0

 ### 내 답안 C-3

X =  1_액션 2_로맨스 3_스릴러 4_다큐멘터리에에 "exists" "count" "ratio"
y =  0 1 2 3
feature =  4*("exists" "count" "ratio") 피처 12개개
target =  븐류, 0 1 2 3

이 숫자는 크기/거리 의미가 있는가? = 존재는 바이노미얼(분류) / 카운트는 빈도수(수치및연속) 비율은 상호 관련성(수치,연속-비율척도)  
class 이름을 숫자로 바꾼 것인가? =  
문제 유형 =  분류문제제
회귀로 풀면 생기는 문제 =  연속형-즉 수치형으로 아웃풋이 안나오고 , 인테거가 아니면 그 결정계수에 따라 정해주고ㅡ 이에 대한 분류로 회귀되어야함. 

근거:
-
 feature_1_액션  feature_2_로맨스  feature_3_스릴러  feature_4_다큐멘터리
0         0.125            0.0            0.0         0.000000
1         0.000            0.1            0.0         0.000000
2         0.000            0.0            0.1         0.000000
3         0.000            0.0            0.0         0.111111
   액션_exists  액션_count  액션_ratio  로맨스_exists  로맨스_count  로맨스_ratio  \
0        1.0       1.0     0.125         0.0        0.0        0.0   
1        0.0       0.0     0.000         1.0        1.0        0.1   
2        0.0       0.0     0.000         0.0        0.0        0.0   
3        0.0       0.0     0.000         0.0        0.0        0.0   

   스릴러_exists  스릴러_count  스릴러_ratio  다큐멘터리_exists  다큐멘터리_count  다큐멘터리_ratio  
0         0.0        0.0        0.0           0.0          0.0     0.000000  
1         0.0        0.0        0.0           0.0          0.0     0.000000  
2         1.0        1.0        0.1           0.0          0.0     0.000000  
3         0.0        0.0        0.0           1.0          1.0     0.111111  
target 이름: ['액션' '로맨스' '스릴러' '다큐멘터리']
y 고유값: [0 1 2 3]
클래스별 개수:
genre_code
0    1
1    1
2    1
3    1
Name: count, dtype: int64
...
0       특수부대 요원이 폭발과 총격 액션 속에서 테러리스트를 추격한다.           0
1    두 사람이 우연히 만나 사랑과 로맨스를 키우지만 이별 위기에 놓인다.           1
2   연쇄 살인범을 쫓는 형사가 어두운 도시에서 스릴러 같은 긴장을 겪는다.           2
3  북극곰과 기후 변화를 담은 다큐멘터리 영화가 자연의 아름다움을 보여준다.           3
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...

## C-5. 교통 혼잡도 class

### 문제

교통 혼잡도를 다음처럼 분류한다.

- 0 = 원활
- 1 = 보통
- 2 = 혼잡
- 3 = 매우 혼잡

시간대, 요일, 날씨, 차량 수, 평균 속도, 사고 여부 등의 정보를 이용해 교통 혼잡도를 예측한다.

### 답할 것

- X =
- y =
- feature =
- target =
- 이 숫자는 크기/거리 의미가 있는가?
- class 이름을 숫자로 바꾼 것인가?
- 문제 유형 =
- 회귀로 풀면 생기는 문제 =
- 단순 다중분류와 순서형 분류 중 무엇이 더 자연스러운가?

In [ ]:
import pandas as pd
import numpy as np

target_names = ["원활", "보통", "혼잡", "매우혼잡"]

#피처1 : 차량수 / 도로용량 비율
#피처2 : 시간대별 가중치/요일별 가중치 
#피처3 : 평균속도/도로구간 비율
#피처4 : 날씨별 가중치 
#피처5 : 사고여부/도로구간 비율
#피처6 : 차량수 / 시간대별 기대용량 비율

## 피처 정의표 (Gate 답안용)
feature_spec = pd.DataFrame({
    "피처": ["feature_1", "feature_2", "feature_3", "feature_4", "feature_5", "feature_6"],
    "한글명": [
        "차량수/도로용량 비율",
        "시간대가중치/요일가중치",
        "평균속도/도로길이 비율",
        "날씨 가중치",
        "사고여부/도로길이 비율",
        "차량수/시간대용량 비율",
    ],
    "수식": [
        "vehicle_count / road_capacity",
        "time_weight[hour] / day_weight[dayofweek]",
        "avg_speed / road_km",
        "weather_weight[weather]",
        "accident / road_km",
        "vehicle_count / hour_capacity[hour]",
    ],
    "타입": ["비율", "비율", "비율", "가중치", "비율", "비율"],
})


# 보조 매핑 (로데이터를 사람이 읽게)
day_names = ["월", "화", "수", "목", "금", "토", "일"]
hour_bucket = {
    range(0, 6): "새벽",
    range(6, 10): "출근",
    range(10, 17): "낮",
    range(17, 20): "퇴근",
    range(20, 24): "야간",
}
def get_hour_bucket(h):
    for r, name in hour_bucket.items():
        if h in r:
            return name
    return "기타"

# 1) 로데이터 직접 제작
raw_df = pd.DataFrame({
    "vehicle_count": [80, 150, 320, 410, 200, 90, 380, 420],
    "hour":          [3, 10, 8, 18, 14, 6, 19, 8],
    "dayofweek":     [1, 2, 1, 4, 3, 6, 5, 1],
    "weather":       ["맑음", "맑음", "비", "흐림", "맑음", "맑음", "비", "눈"],
    "avg_speed":     [70, 55, 25, 15, 45, 65, 20, 10],
    "road_km":       [3.0, 5.0, 4.0, 6.0, 5.0, 3.0, 4.0, 5.0],
    "road_capacity": [200, 300, 400, 400, 300, 200, 400, 400],
    "accident":      [0, 0, 1, 1, 0, 0, 1, 1],
    "congestion_code": [0, 1, 2, 3, 1, 0, 2, 3],
})


time_weight = {h: 1.0 for h in range(24)}
for h, w in {7:1.5, 8:2.0, 9:1.8, 17:1.9, 18:2.2, 19:1.7}.items():
    time_weight[h] = w

day_weight = {0:1.0, 1:1.0, 2:1.0, 3:1.0, 4:1.0, 5:0.7, 6:0.6}

weather_weight = {"맑음":1.0, "흐림":1.1, "비":1.3, "눈":1.5}

hour_capacity = {h: (300 if h in [7,8,17,18,19] else 200) for h in range(24)}

def extract_traffic_features(row):
    return pd.Series({
        "feature_1_vehicle_capacity_ratio": row["vehicle_count"] / row["road_capacity"],
        "feature_2_time_day_weight_ratio": time_weight[row["hour"]] / day_weight[row["dayofweek"]],
        "feature_3_speed_road_ratio": row["avg_speed"] / row["road_km"],
        "feature_4_weather_weight": weather_weight[row["weather"]],
        "feature_5_accident_road_ratio": row["accident"] / row["road_km"],
        "feature_6_vehicle_hour_ratio": row["vehicle_count"] / hour_capacity[row["hour"]],
    })


# 2) 피처 추출
feature_df = raw_df.apply(extract_traffic_features, axis=1)
X_df = feature_df.copy()
X = X_df.values
y = raw_df["congestion_code"].values
y_s = pd.Series(y, name="congestion_code")

# 통합 테이블: 로데이터 + 해석 컬럼 + 피처 + 정답
df = raw_df.copy()
df["요일"] = df["dayofweek"].map(lambda d: day_names[d])
df["시간대구간"] = df["hour"].map(get_hour_bucket)
df["혼잡도"] = df["congestion_code"].map(dict(enumerate(target_names)))
df = pd.concat([df, X_df], axis=1)

# 클래스 코드 → 한글 라벨 매핑 (rename(index=배열)은 Series에서 오류)
class_counts = y_s.value_counts().sort_index()
class_counts.index = [target_names[i] for i in class_counts.index]

print("=" * 60)
print("【1】 피처 정의")
print(feature_spec.to_string(index=False))

print("\n" + "=" * 60)
print("【2】 로데이터 + 분류 라벨")
print(df[[
    "vehicle_count", "hour", "시간대구간", "dayofweek", "요일",
    "weather", "avg_speed", "road_km", "accident",
    "congestion_code", "혼잡도"
]].to_string(index=True))

print("\n" + "=" * 60)
print("【3】 X / y 크기")
print(f"X shape: {X.shape}  → {X.shape[0]}개 샘플, {X.shape[1]}개 피처")
print(f"y shape: {y.shape}  → 클래스 코드 {list(np.unique(y))}")
print(f"target 이름: {list(target_names)}")

print("\n" + "=" * 60)
print("【4】 클래스별 개수")
print(class_counts)

print("\n" + "=" * 60)
print("【5】 피처 통계")
print(X_df.describe().round(3).to_string())

print("\n" + "=" * 60)
print("【6】 샘플별 한 줄 요약")
for i, row in df.iterrows():
    print(
        f"[{i}] {row['혼잡도']}({row['congestion_code']}) | "
        f"{row['hour']}시({row['시간대구간']}) {row['요일']} | "
        f"차량 {row['vehicle_count']}대 | 속도 {row['avg_speed']}km/h | "
        f"날씨 {row['weather']} | "
        f"f1={row['feature_1_vehicle_capacity_ratio']:.3f}, "
        f"f6={row['feature_6_vehicle_hour_ratio']:.3f}"
    )

print("\n" + "=" * 60)
print("【7】 Gate 판단 요약")
print("X =", list(X_df.columns))
print("y = congestion_code (0~3)")
print("feature = 위 6개 수치 피처 (비율/가중치)")
print("target =", list(target_names), "← class 이름을 숫자로 인코딩")
print("문제 유형 = 순서형 다중분류 (원활 < 보통 < 혼잡 < 매우혼잡)")
print("회귀로 풀면 = 2.3 같은 중간값 예측, 클래스 경계 모호")

df

【1】 피처 정의
       피처          한글명                                        수식  타입
feature_1  차량수/도로용량 비율             vehicle_count / road_capacity  비율
feature_2 시간대가중치/요일가중치 time_weight[hour] / day_weight[dayofweek]  비율
feature_3 평균속도/도로길이 비율                       avg_speed / road_km  비율
feature_4       날씨 가중치                   weather_weight[weather] 가중치
feature_5 사고여부/도로길이 비율                        accident / road_km  비율
feature_6 차량수/시간대용량 비율       vehicle_count / hour_capacity[hour]  비율

【2】 로데이터 + 분류 라벨
   vehicle_count  hour 시간대구간  dayofweek 요일 weather  avg_speed  road_km  accident  congestion_code   혼잡도
0             80     3    새벽          1  화      맑음         70      3.0         0                0    원활
1            150    10     낮          2  수      맑음         55      5.0         0                1    보통
2            320     8    출근          1  화       비         25      4.0         1                2    혼잡
3            410    18    퇴근          4  금      흐림         15      6.0   

,vehicle_count,hour,dayofweek,weather,avg_speed,road_km,road_capacity,accident,congestion_code,요일,시간대구간,혼잡도,feature_1_vehicle_capacity_ratio,feature_2_time_day_weight_ratio,feature_3_speed_road_ratio,feature_4_weather_weight,feature_5_accident_road_ratio,feature_6_vehicle_hour_ratio
0,80,3,1,맑음,70,3.0,200,0,0,화,새벽,원활,0.400000,1.000000,23.333333,1.0,0.000000,0.400000
1,150,10,2,맑음,55,5.0,300,0,1,수,낮,보통,0.500000,1.000000,11.000000,1.0,0.000000,0.750000
2,320,8,1,비,25,4.0,400,1,2,화,출근,혼잡,0.800000,2.000000,6.250000,1.3,0.250000,1.066667
3,410,18,4,흐림,15,6.0,400,1,3,금,퇴근,매우혼잡,1.025000,2.200000,2.500000,1.1,0.166667,1.366667
4,200,14,3,맑음,45,5.0,300,0,1,목,낮,보통,0.666667,1.000000,9.000000,1.0,0.000000,1.000000
5,90,6,6,맑음,65,3.0,200,0,0,일,출근,원활,0.450000,1.666667,21.666667,1.0,0.000000,0.450000
6,380,19,5,비,20,4.0,400,1,2,토,퇴근,혼잡,0.950000,2.428571,5.000000,1.3,0.250000,1.266667
7,420,8,1,눈,10,5.0,400,1,3,화,출근,매우혼잡,1.050000,2.000000,2.000000,1.5,0.200000,1.400000


# 코드 힌트:
# X 후보:
# 시간대, 요일, 날씨, 차량 수, 평균 속도, 사고 여부, 도로 구간
# y 후보:
# congestion_level

# 생각할 것:
# 0,1,2,3 사이에는 순서가 있는가?
# 그러나 0과 1의 차이, 2와 3의 차이가 정확히 같은 거리인가?
# 단순 class인가, ordinal class인가?

### 내 답안 C-5

X =  
y =  
feature =  
target =  

이 숫자는 크기/거리 의미가 있는가? =  feature_2 시간대가중치/요일가중치 time_weight[hour] / day_weight[dayofweek]  비율 
및 feature_4       날씨 가중치                   weather_weight[weather] 가중치 에 대해서 구체적 파라미터 수정필요. 

class 이름을 숫자로 바꾼 것인가? =  
문제 유형 =  분류문제
회귀로 풀면 생기는 문제 =  
단순 다중분류와 순서형 분류 중 무엇이 더 자연스러운가? =  

근거:
-
============================================================
【1】 피처 정의
       피처          한글명                                        수식  타입
feature_1  차량수/도로용량 비율             vehicle_count / road_capacity  비율
feature_2 시간대가중치/요일가중치 time_weight[hour] / day_weight[dayofweek]  비율
feature_3 평균속도/도로길이 비율                       avg_speed / road_km  비율
feature_4       날씨 가중치                   weather_weight[weather] 가중치
feature_5 사고여부/도로길이 비율                        accident / road_km  비율
feature_6 차량수/시간대용량 비율       vehicle_count / hour_capacity[hour]  비율

============================================================
【2】 로데이터 + 분류 라벨
   vehicle_count  hour 시간대구간  dayofweek 요일 weather  avg_speed  road_km  accident  congestion_code   혼잡도
0             80     3    새벽          1  화      맑음         70      3.0         0                0    원활
1            150    10     낮          2  수      맑음         55      5.0         0                1    보통
2            320     8    출근          1  화       비         25      4.0         1                2    혼잡
3            410    18    퇴근          4  금      흐림         15      6.0         1                3  매우혼잡
4            200    14     낮          3  목      맑음         45      5.0         0                1    보통
5             90     6    출근          6  일      맑음         65      3.0         0                0    원활
6            380    19    퇴근          5  토       비         20      4.0         1                2    혼잡
7            420     8    출근          1  화       눈         10      5.0         1                3  매우혼잡

============================================================
【3】 X / y 크기
X shape: (8, 6)  → 8개 샘플, 6개 피처
...
feature = 위 6개 수치 피처 (비율/가중치)
target = ['원활', '보통', '혼잡', '매우혼잡'] ← class 이름을 숫자로 인코딩
문제 유형 = 순서형 다중분류 (원활 < 보통 < 혼잡 < 매우혼잡)
회귀로 풀면 = 2.3 같은 중간값 예측, 클래스 경계 모호
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
vehicle_count	hour	dayofweek	weather	avg_speed	road_km	road_capacity	accident	congestion_code	요일	시간대구간	혼잡도	feature_1_vehicle_capacity_ratio	feature_2_time_day_weight_ratio	feature_3_speed_road_ratio	feature_4_weather_weight	feature_5_accident_road_ratio	feature_6_vehicle_hour_ratio
0	80	3	1	맑음	70	3.0	200	0	0	화	새벽	원활	0.400000	1.000000	23.333333	1.0	0.000000	0.400000
1	150	10	2	맑음	55	5.0	300	0	1	수	낮	보통	0.500000	1.000000	11.000000	1.0	0.000000	0.750000
2	320	8	1	비	25	4.0	400	1	2	화	출근	혼잡	0.800000	2.000000	6.250000	1.3	0.250000	1.066667
3	410	18	4	흐림	15	6.0	400	1	3	금	퇴근	매우혼잡	1.025000	2.200000	2.500000	1.1	0.166667	1.366667
4	200	14	3	맑음	45	5.0	300	0	1	목	낮	보통	0.666667	1.000000	9.000000	1.0	0.000000	1.000000
5	90	6	6	맑음	65	3.0	200	0	0	일	출근	원활	0.450000	1.666667	21.666667	1.0	0.000000	0.450000
6	380	19	5	비	20	4.0	400	1	2	토	퇴근	혼잡	0.950000	2.428571	5.000000	1.3	0.250000	1.266667
7	420	8	1	눈	10	5.0	400	1	3	화	출근	매우혼잡	1.050000	2.000000	2.000000	1.5	0.200000	1.400000

확인했다. 채점 범위는 **C-1, C-2, C-3, C-5**다. **C-4는 일부러 안 푼 것으로 보고 채점 제외**한다.
전체 판정부터 말하면, 너는 **X/y 분리 감각은 올라왔고, 코드로 데이터 구조를 확인하려는 태도도 좋다.** 다만 이번 Drill C의 핵심인 **“숫자로 된 y가 실제 수치량인지, class encoding인지”**를 답할 때 자꾸 **feature의 숫자 성격**을 설명하고 있다. 이게 가장 큰 교정 포인트다.

## 전체 판정

| 항목                | 판정                              |
| ----------------- | ------------------------------- |
| X/y 분리            | 대체로 가능                          |
| feature/target 구분 | 가능하지만 target 설명이 가끔 이름/코드로 섞임   |
| 숫자 label 판단       | 아직 불안정                          |
| 회귀/분류 판단          | 대체로 맞음                          |
| 코드 셀 구성           | 시도는 좋으나 clean run 재현성 문제 있음     |
| 최종 점수             | **65/100: 부분 통과, 핵심 함정 재교정 필요** |

이번 Drill C의 중심 질문은 이것이었다.

```text
y의 숫자가 실제 크기/거리/양을 뜻하는가?
아니면 class 이름을 숫자로 바꾼 것인가?
```

그런데 네 답안은 여러 문항에서 이 질문에 대해 **X feature의 수치성**을 설명했다. 예를 들어 Iris에서 “피처는 연속형 수치”라고 답했는데, 문제는 `sepal length`가 수치형인지 묻는 게 아니라 **y = 0,1,2가 수치량인지 class encoding인지**를 묻는 것이다.

---

## C-1. Iris 품종 encoding

### 네 답안 요약

```text
X = shape(150,4)
y = shape(150,)
feature = sepal/petal 네 개
target = setosa, versicolor, virginica
이 숫자는 크기/거리 의미가 있는가? = X 피처는 연속형 수치...
class 이름을 숫자로 바꾼 것인가? = 클래스 이름을 숫자로 바꾼 게 아님
문제 유형 = 다중분류
```

### 채점

| 항목                | 판정      |
| ----------------- | ------- |
| X                 | 부분 정답   |
| y                 | 부분 정답   |
| feature           | 정답      |
| target            | 정답에 가까움 |
| 숫자 의미 판단          | 오답      |
| class encoding 판단 | 오답      |
| 문제 유형             | 정답      |
| 회귀로 풀면 생기는 문제     | 불충분     |

### 구체 교정

`X = shape(150,4)`라고 쓴 것은 틀리지는 않지만 부족하다. (X)는 단순히 shape가 아니라 **꽃의 네 가지 측정값 전체**다.

더 정확한 표현은 이렇다.

```text
X = sepal length, sepal width, petal length, petal width로 구성된 입력 행렬
X.shape = (150, 4)
```

`y = shape(150,)`도 마찬가지다. shape만 쓰면 target의 의미가 빠진다.

```text
y = species_code, 즉 setosa/versicolor/virginica를 0/1/2로 인코딩한 값
y.shape = (150,)
```

가장 중요한 오답은 여기다.

```text
class 이름을 숫자로 바꾼 것인가? = 클래스 이름을 숫자로 바꾼 게 아님
```

이건 반대로 써야 한다. Iris에서 `0=setosa`, `1=versicolor`, `2=virginica`는 **class 이름을 숫자로 바꾼 것**이다. `0,1,2` 사이에는 크기나 거리 의미가 없다. `virginica`가 `setosa`보다 “2만큼 큰 품종”이라는 뜻이 아니다.

회귀로 풀면 생기는 문제도 더 명확해야 한다. 회귀로 풀면 모델은 `0,1,2` 사이의 거리와 순서가 의미 있다고 가정한다. 예를 들어 예측값이 `1.7`이면 이것을 어떻게 해석할지 애매해진다. 또한 MSE는 `0→2` 오류를 `0→1` 오류보다 더 큰 수치 오차로 취급하는데, 품종 class에 그런 거리 의미가 있다는 보장은 없다.

### C-1 점수

**60/100.**
X/y와 문제 유형은 맞았지만, Drill C의 핵심인 class encoding 판단을 반대로 썼다.

---

## C-2. 고객 만족도 1~5점

### 네 답안 요약

```text
feature = 응답 시간, 재방문 횟수, 불만 키워드 수, 구매 금액, 상담 횟수
target = 매우불만족/불만족/보통/만족/매우만족 또는 1,2,3,4,5
이 숫자는 크기/거리 의미가 있는가? = 단위가 다르며 범주형 아님
class 이름을 숫자로 바꾼 것인가? = 클래스 아이디를 숫자로 변환한 게 아님
문제 유형 = 5중 분류
다른 방식 = 회귀 후 반올림 가능
```

### 채점

| 항목                | 판정      |
| ----------------- | ------- |
| X                 | 누락      |
| y                 | 누락에 가까움 |
| feature           | 정답      |
| target            | 부분 정답   |
| 숫자 의미 판단          | 부분 오답   |
| class encoding 판단 | 부분 정답   |
| 문제 유형             | 부분 정답   |
| 다른 방식 설명          | 괜찮음     |

### 구체 교정

C-2는 애매한 문제다. 이 문제를 잘 푸는 방식은 “하나로 단정”이 아니라 **왜 애매한지**를 설명하는 것이다.

고객 만족도 `1,2,3,4,5`는 Iris의 `0,1,2`와 다르다. Iris 숫자는 완전히 명목형 class encoding이다. 하지만 만족도 점수는 보통 다음 순서를 가진다.

```text
1 < 2 < 3 < 4 < 5
```

즉, **순서 의미는 있다.**
하지만 `1점→2점` 차이와 `4점→5점` 차이가 정확히 같은 심리적 거리인지 확실하지 않다. 그래서 이것은 보통 **순서형 target**이다.

가능한 접근은 세 가지다.

```text
1. 다중분류: 1,2,3,4,5를 다섯 class로 본다.
2. 순서형 분류: 1<2<3<4<5 순서를 반영한다.
3. 회귀: 1~5를 거의 연속 점수처럼 보고 예측 후 반올림한다.
```

네가 “5중 분류”라고 한 것은 완전히 틀리지는 않다. 하지만 가장 좋은 답은 **순서형 분류 또는 ordinal target**이라고 쓰는 것이다.

“이 숫자는 크기/거리 의미가 있는가?”에 대한 네 답은 질문을 잘못 잡았다. 여기서 묻는 숫자는 feature의 단위가 아니라 **y = 1~5 만족도 점수**다.

더 정확히는:

```text
이 숫자는 순서 의미는 있다.
다만 정확한 거리 의미는 약하다.
```

“class 이름을 숫자로 바꾼 것인가?”도 애매하다. 만약 실제 설문이 `매우불만족/불만족/보통/만족/매우만족`을 1~5로 저장한 것이라면 class 이름을 숫자로 매핑한 것이다. 다만 완전한 명목형 class encoding이 아니라 **순서형 encoding**이다.

### C-2 점수

**70/100.**
문제 유형을 분류로 잡은 것은 가능하지만, “순서형”이라는 핵심 표현이 빠졌고 숫자 의미 판단이 흔들렸다.

---

## C-3. 영화 장르 encoding

### 네 답안 요약

```text
X = 액션/로맨스/스릴러/다큐멘터리의 exists/count/ratio
y = 0 1 2 3
feature = 12개
target = 분류, 0 1 2 3
이 숫자는 크기/거리 의미가 있는가? = exists/count/ratio 설명
class 이름을 숫자로 바꾼 것인가? = 공란
문제 유형 = 분류
회귀로 풀면 생기는 문제 = 연속형 출력, integer가 아니면 결정계수로 정해야 함
```

### 채점

| 항목                | 판정    |
| ----------------- | ----- |
| X                 | 부분 정답 |
| y                 | 정답    |
| feature           | 부분 정답 |
| target            | 부분 정답 |
| 숫자 의미 판단          | 오답    |
| class encoding 판단 | 누락    |
| 문제 유형             | 정답    |
| 회귀 문제 설명          | 부분 정답 |

### 구체 교정

코드 자체는 좋은 시도다. 텍스트를 그대로 모델에 넣지 않고 `exists`, `count`, `ratio` 같은 수치 feature로 바꾼 것은 Gate 0 관점에서 적절하다.

다만 네 코드에서는 실제 (X)로 무엇을 썼는지 답안이 조금 불일치한다. 코드에서는 최종 `X_df`를 ratio 4개로 만들었다.

```text
feature_1_액션
feature_2_로맨스
feature_3_스릴러
feature_4_다큐멘터리
```

그런데 답안에는 exists/count/ratio 12개라고 썼다. 둘 중 하나로 고정해야 한다.

가능한 답안은 둘 중 하나다.

```text
버전 A:
X = 네 장르 키워드 ratio 4개
feature = 액션_ratio, 로맨스_ratio, 스릴러_ratio, 다큐멘터리_ratio
```

또는

```text
버전 B:
X = 네 장르 키워드의 exists/count/ratio 12개
feature = 액션_exists, 액션_count, 액션_ratio, ... 총 12개
```

이번 문제의 핵심 오답은 또 같은 지점이다.

```text
이 숫자는 크기/거리 의미가 있는가?
```

여기서 “이 숫자”는 feature의 exists/count/ratio가 아니라 **y = 0,1,2,3 장르 코드**다.

정확한 판단은 다음이다.

```text
0,1,2,3은 크기/거리 의미가 없다.
0=액션, 1=로맨스, 2=스릴러, 3=다큐멘터리라는 class 이름을 숫자로 바꾼 것이다.
```

영화 장르에 순서나 거리 의미는 없다. 액션이 로맨스보다 작거나, 다큐멘터리가 스릴러보다 큰 것이 아니다. 따라서 이것은 **명목형 다중분류**다.

회귀로 풀면 생기는 문제는 다음처럼 써야 한다.

```text
회귀로 풀면 장르 코드 사이에 순서와 거리 의미가 있다고 가정한다.
예측값이 1.6처럼 나오면 어떤 장르인지 애매하다.
또한 MSE가 액션(0)과 다큐(3)의 차이를 액션(0)과 로맨스(1)의 차이보다 더 큰 오류로 처리하는데,
장르 class에는 그런 수치적 거리 의미가 없다.
```

“결정계수”라는 표현은 여기서 부적절하다. 결정계수 (R^2)는 회귀 평가 지표이고, class를 정하는 기준이 아니다. 여기서는 “threshold” 또는 “rounding”이라고 해야 한다.

### C-3 점수

**68/100.**
텍스트 feature engineering 시도는 좋지만, target code의 의미 판단이 빠졌고 회귀 문제 설명에 용어 오류가 있다.

---

## C-4. 시험 점수 예측

채점 제외.
하지만 이 문제는 반드시 나중에 풀어야 한다. 이유는 C-1, C-3과 반대되는 **진짜 회귀 예시**이기 때문이다.

Iris의 `0,1,2`는 class encoding이지만, 시험 점수 `0~100`은 실제 수치량이다. `80점`과 `90점`의 차이는 10점이고, `50점`과 `60점`의 차이도 10점이다. 따라서 기본적으로 회귀다.

이 문항을 안 풀면 “숫자로 된 y는 대부분 분류인가?”라는 반대 오해가 생긴다. 다음에 반드시 포함한다.

---

## C-5. 교통 혼잡도 class

### 네 답안 요약

코드에서는 feature를 꽤 잘 만들었다.

```text
feature_1 = 차량수/도로용량 비율
feature_2 = 시간대가중치/요일가중치
feature_3 = 평균속도/도로길이 비율
feature_4 = 날씨 가중치
feature_5 = 사고여부/도로길이 비율
feature_6 = 차량수/시간대용량 비율
y = congestion_code
target 이름 = 원활, 보통, 혼잡, 매우혼잡
```

그런데 답안 작성 칸에서는 X, y, feature, target이 비어 있고, “이 숫자는 크기/거리 의미가 있는가?”에 feature_2, feature_4 설명을 넣었다.

### 채점

| 항목                | 판정                   |
| ----------------- | -------------------- |
| X                 | 코드상 정답, 답안칸 누락       |
| y                 | 코드상 정답, 답안칸 누락       |
| feature           | 코드상 정답, 답안칸 누락       |
| target            | 코드상 정답, 답안칸 누락       |
| 숫자 의미 판단          | 답안칸 기준 오답            |
| class encoding 판단 | 누락                   |
| 문제 유형             | 부분 정답                |
| 순서형 판단            | 출력 근거에는 있음, 답안칸에는 누락 |

### 구체 교정

C-5는 코드 설계가 가장 좋았다. 실제로 (X)를 6개 feature로 만들고, (y)를 `congestion_code`로 만든 것은 적절하다. 다만 최종 답안칸이 비어 있어서 채점상 감점이 크다. 노트북에서는 출력 로그를 길게 붙이는 것보다, 마지막 Markdown에 네 판단을 압축해서 써야 한다.

정확한 답안 구조는 이렇다.

```text
X = feature_1~feature_6으로 구성된 수치 feature matrix
y = congestion_code
feature = 차량수/도로용량 비율, 시간대/요일 가중치, 평균속도/도로길이 비율, 날씨 가중치, 사고여부/도로길이 비율, 차량수/시간대용량 비율
target = 원활, 보통, 혼잡, 매우혼잡
```

그리고 핵심 질문에 대한 답은 이렇다.

```text
이 숫자는 크기/거리 의미가 있는가?
= 순서 의미는 있다. 0<1<2<3으로 혼잡도가 커진다.
  그러나 0과 1의 차이, 2와 3의 차이가 정확히 같은 거리라고 단정하기는 어렵다.
```

```text
class 이름을 숫자로 바꾼 것인가?
= 맞다. 원활/보통/혼잡/매우혼잡을 0/1/2/3으로 인코딩한 것이다.
  단, Iris나 영화 장르와 달리 순서가 있는 class encoding이다.
```

문제 유형은 “분류문제”라고만 쓰기보다 더 정확히는:

```text
문제 유형 = 순서형 다중분류
```

회귀로 풀면 생기는 문제:

```text
회귀로 풀면 2.3 같은 중간값이 나올 수 있다.
또 MSE는 0→3 오류를 0→1 오류보다 훨씬 크게 보는데,
혼잡도 등급 사이의 간격이 실제로 동일하다고 보장되지 않는다.
다만 혼잡도가 순서형이므로 ordinal regression 또는 회귀 후 thresholding 같은 접근은 보조적으로 가능하다.
```

### 추가 보완: feature engineering 관련

네가 feature_2, feature_4에 대해 “구체적 파라미터 수정 필요”라고 쓴 것은 좋은 감각이다. 다만 표현을 이렇게 바꾸는 게 좋다.

```text
feature_2와 feature_4는 raw data가 아니라 사람이 설계한 가중치 기반 feature다.
이 가중치가 도메인 지식으로 사전에 정해진 것인지,
전체 데이터의 target을 보고 조정한 것인지 구분해야 한다.
target을 보고 조정했다면 leakage 위험이 있다.
```

즉, `weather_weight`, `time_weight`는 그냥 feature가 아니라 **feature engineering 규칙**이다. 이 규칙이 test 데이터 정보를 보고 만들어지면 누수가 된다.

### C-5 점수

**70/100.**
코드와 feature 구성은 좋지만, 최종 답안칸이 비어 있고 핵심 질문을 feature 설명으로 답했다. 판단 자체는 근거 출력 부분에 일부 들어 있으므로 부분 가산한다.

---

## 코드 셀 관련 피드백

### 1. C-1 코드 재현성 문제

C-1 코드 셀은 출력은 잘 나왔지만, 소스만 보면 clean kernel에서 실패할 가능성이 있다. `X`, `y`, `X_df`, `y_s`를 만들기 전에 `X_df.head(10)`을 호출하고 있다.

네 노트북이 실행된 이유는 아마 이전 커널 상태에 변수가 남아 있었기 때문이다. 노트북은 항상 **Kernel Restart & Run All** 기준으로 재현되어야 한다.

보완 힌트:

```text
iris.data를 X에 넣는다.
iris.target을 y에 넣는다.
iris.data와 iris.feature_names로 X_df를 만든다.
iris.target으로 y_s를 만든다.
그 다음 head/info/describe를 확인한다.
```

코드 전체를 외울 필요는 없다. 중요한 것은 순서다.

```text
load → X/y 정의 → DataFrame/Series 생성 → shape/info 확인
```

### 2. C-2 코드 셀

`from sklearn.datasets import load_만족도`는 실제 존재하지 않는다. 네가 “예상 코드”로 쓴 것은 이해하지만, 노트북에서는 실행할 수 없는 코드와 실행 가능한 코드를 구분해야 한다.

이럴 때는 code cell에 실제 실행 코드 대신 Markdown으로 “가상 데이터 구조”를 쓰거나, 아주 작은 예시 DataFrame을 직접 만드는 방식이 낫다.

### 3. C-3 코드 셀

C-3는 좋은 시도다. 텍스트를 숫자 feature로 바꾸는 방향이 맞다. 다만 너무 많은 출력이 Markdown 답안에 그대로 붙었다. 노트북에서는 출력과 판단을 분리해야 한다.

권장 구조:

```text
Code cell = 데이터 구조 확인
Markdown cell = 내가 판단한 X/y/type 정리
```

### 4. C-5 코드 셀

C-5는 feature engineering 감각이 좋다. 특히 차량수/용량, 평균속도, 사고 여부, 날씨 가중치 등을 feature로 만든 것은 실제 모델링 사고에 가깝다.

다만 Gate 0-1에서는 코드 복잡도보다 **target code 해석**이 우선이다. 지금은 feature engineering 쪽으로 너무 많이 갔다. 다음 답안에서는 코드보다 다음 네 줄을 정확히 쓰는 것이 더 중요하다.

```text
y = congestion_code
0/1/2/3은 class encoding이다.
단, 순서가 있다.
정확한 거리 의미는 약하다.
```



## 추가 보완 문제

다음에는 바로 이 6개만 풀면 된다. 답안은 짧게, 코드 없이 Markdown만 쓰면 된다.

### 보완 1

`0=남성, 1=여성`으로 저장된 성별을 예측한다.

```text
이 숫자는 크기/거리 의미가 있는가?
class encoding인가?
문제 유형은?
회귀로 풀면 무엇이 문제인가?
```

### 보완 2

`0=정상, 1=주의, 2=위험, 3=매우위험`으로 저장된 안전 등급을 예측한다.

```text
순서 의미가 있는가?
거리 의미가 있는가?
문제 유형은?
단순 다중분류와 순서형 분류 중 무엇이 더 적절한가?
```

### 보완 3

학생의 수학 점수를 0~100점으로 예측한다.

```text
이 숫자는 실제 수치량인가?
문제 유형은?
분류로 바꾸려면 y를 어떻게 재정의해야 하는가?
```

### 보완 4

고객의 연령대를 `0=10대, 1=20대, 2=30대, 3=40대`로 예측한다.

```text
순서 의미가 있는가?
거리 의미가 있는가?
회귀인가, 분류인가?
```

### 보완 5

상품 카테고리를 `0=전자제품, 1=의류, 2=식품, 3=가구`로 예측한다.

```text
class encoding인가?
숫자 크기 의미가 있는가?
문제 유형은?
```

### 보완 6

기온을 섭씨 단위로 예측한다.

```text
이 숫자는 실제 수치량인가?
문제 유형은?
분류로 바꾸려면 어떻게 해야 하는가?
```

이번 보완은 코드 금지다.
목표는 구현이 아니라 **y의 의미 판별**이다.


보완 1 단순 명목형 분류문제이므로 상호 0/1 이진분류 (바이노미널)이다. 따라서 회귀문제로 접근시ㅡ 수치연속형으로 출력이되는데, 1) 클래스 인코딩으로 실제 수치가 아니며, 또한 0.5 0.7 등 상호간 거리성에 대해판별 또는 예측 , 인사이트를 얻고자 함이 아니므로 회귀로 풀면 안된다.

보완 2 안전등급에 대한 순서형 즉 오디널 클래시피케이션이다. 순서형으로서 상호 0<1<2<3 에 대하여ㅡ 분류된 것에 대해 가중을 매길 순 있으나, 0<->1 과 2<->3 간 거리에 대해 등간을 적용할 순 없다. 따라서, 회귀형으로 볼시 거리 및 가중에대한 문제가 발생하며, 또한 이는 실제 수치가 아닌 클래스인코딩으로서 거리에 관해 묻는 문제가 아님으로 명확히 분류로 인식해야한다 

보완3 학생의 수학점수를 0~100 으로 예측할 시, 이는 점수에 대해서 구체적 연속값으로서 예측하는 회귀문제이다. 이때, 타겟값은 실제수치이며 클래스 인코딩이 아니다. 이때, 분류문제로 변환하기 위해서는 타겟값 연속 수치형에 대해서 비율 또는 임의로 사전에 급간을 두고, 이에 대해 급간을 클래스 인코딩하여 분류문제로 변환할 수 있다. 예를들면, 9등급제가 있곘다. 이는 앞서 2번처럼 단순 클래스인코딩이 아닌 실 타겟값에 대한 순서형 인코딩으로서 2와 달리 거리가 의미가 있다. 잠시만 이게 오디날클래시피케이션인가?

보완 4 5 는 패스

보완 6은, 기온이란 단위는 절대영도 즉 상대적 0이 아닌 실제 0을 가지고 있으며 음수값으로 표현될 수 있다. 따라서 실제 수치값이자 수치형 범주이며, 비율척도 등간척도가 아니다. 따라서, 이는 회귀문제로 풀수 있으나 이때 음수에 대해 아웃풋이 나올 수 있도록 세부조정은 해야한다. (기존 실제값-예측값)^2 에 대해 진행하는 회귀형 분류수식에 대한 수정을 말함. ) 이때 분류로 변경시에는, 3과 같이 급간으로 나누어 클래스인코딩시 가능한데, 이때 오디널 클래스피케이션으로 클래스 인코딩간 가중을 설정해줄 수 있으며 명확하게 수치형에 기반한 것이므로 3과 같다.